<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Sequencing Sperm Genomes Reveals Reduced Recombination Frequency among Asthenospermia Patients - 4&nbsp; Prepare the input data for HAPI</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./section_4_run_Hapi.html" rel="next">
<link href="./section_2_aneuploidy_test.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./section_3_prepare_data_for_Hapi.html"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Prepare the input data for HAPI</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Sequencing Sperm Genomes Reveals Reduced Recombination Frequency among Asthenospermia Patients</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./section_1_data_process.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Data Preparation</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./section_2_aneuploidy_test.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Aneuploidy Test</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./section_3_prepare_data_for_Hapi.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Prepare the input data for HAPI</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./section_4_run_Hapi.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Run HAPI</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./section_5_filter_HAPI_output.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Filter Hapi Result</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./section_6_compare_CO.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Compare CO</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./section_7_CO_and_clinical_outcome_association_study.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Crossover Frequency and Clinical Characteristics Association Study</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./section_8_CO_whole_genome_association_study.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Crossover Frequency and Genome Association Study</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./section_9_transcriptome.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Sperm Transcriptome Study</span></span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#prepare-the-aneuploidy-removed-sperm-name-list" id="toc-prepare-the-aneuploidy-removed-sperm-name-list" class="nav-link active" data-scroll-target="#prepare-the-aneuploidy-removed-sperm-name-list"><span class="header-section-number">4.1</span> Prepare the aneuploidy removed sperm name list</a></li>
  <li><a href="#generate-the-hpai-input-matrix" id="toc-generate-the-hpai-input-matrix" class="nav-link" data-scroll-target="#generate-the-hpai-input-matrix"><span class="header-section-number">4.2</span> Generate the Hpai input matrix</a></li>
  <li><a href="#remove-the-aneuploidy-sperms" id="toc-remove-the-aneuploidy-sperms" class="nav-link" data-scroll-target="#remove-the-aneuploidy-sperms"><span class="header-section-number">4.3</span> Remove the aneuploidy sperms</a></li>
  <li><a href="#remove-the-aneuploidy" id="toc-remove-the-aneuploidy" class="nav-link" data-scroll-target="#remove-the-aneuploidy"><span class="header-section-number">4.4</span> Remove the aneuploidy</a></li>
  <li><a href="#check-the-na-density" id="toc-check-the-na-density" class="nav-link" data-scroll-target="#check-the-na-density"><span class="header-section-number">4.5</span> Check the NA density</a></li>
  <li><a href="#regenerate-the-hapi-input" id="toc-regenerate-the-hapi-input" class="nav-link" data-scroll-target="#regenerate-the-hapi-input"><span class="header-section-number">4.6</span> Regenerate the Hapi input</a></li>
  <li><a href="#count-the-sperm-number-in-each-individual-after-filter" id="toc-count-the-sperm-number-in-each-individual-after-filter" class="nav-link" data-scroll-target="#count-the-sperm-number-in-each-individual-after-filter"><span class="header-section-number">4.7</span> Count the sperm number in each individual after filter</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Prepare the input data for HAPI</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p>Followed by the last Chapter, we get the filtered SNP positions and next we need to make the sperm VCF files into a R matrix for Hapi. Since genotype phasing and crossover detect only needs heterzygous loci as input, we use the blood sample to get this information then combine with the corresponding positions in sperm to get the matrix.</p>
<section id="prepare-the-aneuploidy-removed-sperm-name-list" class="level2" data-number="4.1">
<h2 data-number="4.1" class="anchored" data-anchor-id="prepare-the-aneuploidy-removed-sperm-name-list"><span class="header-section-number">4.1</span> Prepare the aneuploidy removed sperm name list</h2>
</section>
<section id="generate-the-hpai-input-matrix" class="level2" data-number="4.2">
<h2 data-number="4.2" class="anchored" data-anchor-id="generate-the-hpai-input-matrix"><span class="header-section-number">4.2</span> Generate the Hpai input matrix</h2>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R script</strong></pre>
</div>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>chr_id <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>blood_gt_chr <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(<span class="st">"../datasets/simplified_blood_gt_by_chr/simplified_blood_gt_by_chr"</span>, chr_id, <span class="st">".rds"</span>))</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>SPERM_DIRECTORY <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"../datasets/sperm_vcf_file/selected_vcf/QUAL_and_RGQ/chr"</span>, chr_id, <span class="st">"/"</span>)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>sperm_list <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="at">path =</span> SPERM_DIRECTORY, <span class="at">pattern =</span> <span class="st">"*.vcf.gz$"</span>, <span class="at">all.files =</span> <span class="cn">FALSE</span>, </span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>                              <span class="at">full.names =</span> <span class="cn">FALSE</span>, <span class="at">recursive =</span> <span class="cn">FALSE</span>,</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>                              <span class="at">ignore.case =</span> <span class="cn">FALSE</span>, <span class="at">include.dirs =</span> <span class="cn">FALSE</span>, <span class="at">no.. =</span> <span class="cn">FALSE</span>)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>sperm_list_df <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"</span><span class="sc">\\</span><span class="st">.vcf</span><span class="sc">\\</span><span class="st">.gz"</span>, <span class="st">""</span>, sperm_list)  </span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>sperm_list_df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(sperm_list_df)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="co">#write.table(sperm_list_df, </span></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="co">#            file = "D:/sperm_project/sperm_vcf_file/sperm_sampleList/sperm_sample_nameList.tsv", </span></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="co">#            append = F, sep = "\t", row.names=FALSE, col.names=FALSE, quote=FALSE)</span></span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a><span class="do">## Merge to the matrix</span></span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a>patient_name <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"s</span><span class="sc">\\</span><span class="st">d</span><span class="sc">\\</span><span class="st">.vcf</span><span class="sc">\\</span><span class="st">.gz"</span>, <span class="st">""</span>, sperm_list)</span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a>patient_name <span class="ot">&lt;-</span> <span class="fu">unique</span>(patient_name)</span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (chr_id <span class="cf">in</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">22</span>)) {</span>
<span id="cb1-27"><a href="#cb1-27" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Print the current working chrosome</span></span>
<span id="cb1-28"><a href="#cb1-28" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb1-29"><a href="#cb1-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">"Chromosome:"</span>)</span>
<span id="cb1-30"><a href="#cb1-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(chr_id)</span>
<span id="cb1-31"><a href="#cb1-31" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb1-32"><a href="#cb1-32" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Read the blood genotype by chromosome</span></span>
<span id="cb1-33"><a href="#cb1-33" aria-hidden="true" tabindex="-1"></a>  blood_gt_chr <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(<span class="st">"../datasets/simplified_blood_gt_by_chr/simplified_blood_gt_by_chr"</span>, chr_id, <span class="st">".rds"</span>))</span>
<span id="cb1-34"><a href="#cb1-34" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb1-35"><a href="#cb1-35" aria-hidden="true" tabindex="-1"></a>  <span class="do">## The Directory of filter sperm vcf file</span></span>
<span id="cb1-36"><a href="#cb1-36" aria-hidden="true" tabindex="-1"></a>  Filtered_SPERM_DIRECTORY <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"../datasets/sperm_vcf_file/selected_vcf/QUAL_and_RGQ/chr"</span>, chr_id, <span class="st">"/"</span>)</span>
<span id="cb1-37"><a href="#cb1-37" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb1-38"><a href="#cb1-38" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Initate the Hapi Input Matirx List</span></span>
<span id="cb1-39"><a href="#cb1-39" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb1-40"><a href="#cb1-40" aria-hidden="true" tabindex="-1"></a>  Hapi_input_matrix_List <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb1-41"><a href="#cb1-41" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb1-42"><a href="#cb1-42" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (patient <span class="cf">in</span> patient_name) {</span>
<span id="cb1-43"><a href="#cb1-43" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-44"><a href="#cb1-44" aria-hidden="true" tabindex="-1"></a>    patient_sperm <span class="ot">&lt;-</span> sperm_list[<span class="fu">grepl</span>(<span class="fu">paste0</span>(patient, <span class="st">"s</span><span class="sc">\\</span><span class="st">d"</span>), sperm_list)]</span>
<span id="cb1-45"><a href="#cb1-45" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-46"><a href="#cb1-46" aria-hidden="true" tabindex="-1"></a>    sperm_vcf_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb1-47"><a href="#cb1-47" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-48"><a href="#cb1-48" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (sperm_vcf_file <span class="cf">in</span> patient_sperm) {</span>
<span id="cb1-49"><a href="#cb1-49" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb1-50"><a href="#cb1-50" aria-hidden="true" tabindex="-1"></a>      vcf_file <span class="ot">&lt;-</span> <span class="fu">read.vcfR</span>(<span class="fu">paste0</span>(Filtered_SPERM_DIRECTORY, sperm_vcf_file), <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb1-51"><a href="#cb1-51" aria-hidden="true" tabindex="-1"></a>      vcf_file_fix_gt <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">extract.gt</span>(vcf_file, <span class="at">as.numeric =</span> T, <span class="at">IDtoRowNames =</span> T)) <span class="sc">%&gt;%</span> tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(<span class="st">"POS"</span>)</span>
<span id="cb1-52"><a href="#cb1-52" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb1-53"><a href="#cb1-53" aria-hidden="true" tabindex="-1"></a>      sperm_id <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"</span><span class="sc">\\</span><span class="st">.vcf</span><span class="sc">\\</span><span class="st">.gz"</span>, <span class="st">""</span>, sperm_vcf_file)</span>
<span id="cb1-54"><a href="#cb1-54" aria-hidden="true" tabindex="-1"></a>      sperm_vcf_list[[sperm_id]] <span class="ot">&lt;-</span> vcf_file_fix_gt</span>
<span id="cb1-55"><a href="#cb1-55" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb1-56"><a href="#cb1-56" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-57"><a href="#cb1-57" aria-hidden="true" tabindex="-1"></a>    sperm_matrix <span class="ot">&lt;-</span> sperm_vcf_list <span class="sc">%&gt;%</span> <span class="fu">reduce</span>(full_join, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"POS"</span>))</span>
<span id="cb1-58"><a href="#cb1-58" aria-hidden="true" tabindex="-1"></a>    sperm_matrix <span class="ot">&lt;-</span> sperm_matrix <span class="sc">%&gt;%</span> <span class="fu">separate</span>(POS, <span class="fu">c</span>(<span class="st">'chr'</span>, <span class="st">'pos'</span>))</span>
<span id="cb1-59"><a href="#cb1-59" aria-hidden="true" tabindex="-1"></a>    sperm_matrix <span class="ot">&lt;-</span> sperm_matrix[<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb1-60"><a href="#cb1-60" aria-hidden="true" tabindex="-1"></a>    sperm_matrix <span class="ot">&lt;-</span> sperm_matrix <span class="sc">%&gt;%</span> <span class="fu">mutate_if</span>(is.character, as.integer)</span>
<span id="cb1-61"><a href="#cb1-61" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-62"><a href="#cb1-62" aria-hidden="true" tabindex="-1"></a>    <span class="do">## change NF patient name</span></span>
<span id="cb1-63"><a href="#cb1-63" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-64"><a href="#cb1-64" aria-hidden="true" tabindex="-1"></a>    patient<span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"N"</span>, <span class="st">"NF"</span>, patient)</span>
<span id="cb1-65"><a href="#cb1-65" aria-hidden="true" tabindex="-1"></a>    patient<span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"NFF"</span>, <span class="st">"NF"</span>, patient)</span>
<span id="cb1-66"><a href="#cb1-66" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-67"><a href="#cb1-67" aria-hidden="true" tabindex="-1"></a>    blood_patient_gt <span class="ot">&lt;-</span> blood_gt_chr[ , <span class="fu">c</span>(<span class="st">"chr"</span>, <span class="st">"pos"</span>, <span class="st">"ref"</span>, <span class="st">"alt"</span>, patient)]</span>
<span id="cb1-68"><a href="#cb1-68" aria-hidden="true" tabindex="-1"></a>    blood_patient_het_gt <span class="ot">&lt;-</span> blood_patient_gt[<span class="fu">str_detect</span>(blood_patient_gt[,<span class="dv">5</span>], <span class="st">"0/1|0</span><span class="sc">\\</span><span class="st">|1"</span>), ]</span>
<span id="cb1-69"><a href="#cb1-69" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rownames</span>(blood_patient_het_gt) <span class="ot">&lt;-</span> blood_patient_het_gt<span class="sc">$</span>pos</span>
<span id="cb1-70"><a href="#cb1-70" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-71"><a href="#cb1-71" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rownames</span>(sperm_matrix) <span class="ot">&lt;-</span> sperm_matrix<span class="sc">$</span>pos</span>
<span id="cb1-72"><a href="#cb1-72" aria-hidden="true" tabindex="-1"></a>    common_position <span class="ot">&lt;-</span> <span class="fu">intersect</span>(<span class="fu">rownames</span>(sperm_matrix), <span class="fu">rownames</span>(blood_patient_het_gt))</span>
<span id="cb1-73"><a href="#cb1-73" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-74"><a href="#cb1-74" aria-hidden="true" tabindex="-1"></a>    hapi_input_matrix <span class="ot">&lt;-</span> <span class="fu">cbind</span>(blood_patient_het_gt[common_position, <span class="sc">-</span><span class="dv">5</span>], sperm_matrix[common_position, <span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span>)])</span>
<span id="cb1-75"><a href="#cb1-75" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-76"><a href="#cb1-76" aria-hidden="true" tabindex="-1"></a>    sorted_hapi_input_matrix <span class="ot">&lt;-</span> hapi_input_matrix[<span class="fu">order</span>(hapi_input_matrix<span class="sc">$</span>pos),]</span>
<span id="cb1-77"><a href="#cb1-77" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-78"><a href="#cb1-78" aria-hidden="true" tabindex="-1"></a>    Hapi_input_matrix_List[[patient]] <span class="ot">&lt;-</span> sorted_hapi_input_matrix</span>
<span id="cb1-79"><a href="#cb1-79" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb1-80"><a href="#cb1-80" aria-hidden="true" tabindex="-1"></a>  }  </span>
<span id="cb1-81"><a href="#cb1-81" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb1-82"><a href="#cb1-82" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(Hapi_input_matrix_List, <span class="fu">paste0</span>(<span class="st">"../datasets/Hapi_input/QUAL_and_RGQ/chr"</span>, chr_id, <span class="st">".rds"</span>))</span>
<span id="cb1-83"><a href="#cb1-83" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
<section id="remove-the-aneuploidy-sperms" class="level2" data-number="4.3">
<h2 data-number="4.3" class="anchored" data-anchor-id="remove-the-aneuploidy-sperms"><span class="header-section-number">4.3</span> Remove the aneuploidy sperms</h2>
</section>
<section id="remove-the-aneuploidy" class="level2" data-number="4.4">
<h2 data-number="4.4" class="anchored" data-anchor-id="remove-the-aneuploidy"><span class="header-section-number">4.4</span> Remove the aneuploidy</h2>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R script</strong></pre>
</div>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>aneuploidy_sperm_list <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"AS24s5"</span>, <span class="st">"AS32s4"</span>, <span class="st">"NF25-1s1"</span>, <span class="st">"NF27-3s3"</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>aneuploidy_donor_list <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"AS24"</span>, <span class="st">"AS32"</span>, <span class="st">"NF25-1"</span>, <span class="st">"NF27-3"</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>NA_count_chr_list <span class="ot">&lt;-</span> <span class="fu">list</span>()  </span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (chr_id <span class="cf">in</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">22</span>)) {</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>  Hapi_input <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(<span class="st">"../datasets/Hapi_input/QUAL_and_RGQ/chr"</span>, chr_id, <span class="st">".rds"</span>))</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (donor <span class="cf">in</span> aneuploidy_donor_list) {</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">colnames</span>(Hapi_input[[donor]]) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"N"</span>, <span class="st">"NF"</span>, <span class="fu">colnames</span>(Hapi_input[[donor]]))</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">colnames</span>(Hapi_input[[donor]]) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"NFF"</span>, <span class="st">"NF"</span>, <span class="fu">colnames</span>(Hapi_input[[donor]]))</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>    Hapi_input[[donor]] <span class="ot">&lt;-</span> Hapi_input[[donor]][,<span class="sc">!</span><span class="fu">names</span>(Hapi_input[[donor]]) <span class="sc">%in%</span> aneuploidy_sperm_list]</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>  NA_count_chr <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (donor <span class="cf">in</span> <span class="fu">names</span>(Hapi_input)) {</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">colnames</span>(Hapi_input[[donor]]) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"N"</span>, <span class="st">"NF"</span>, <span class="fu">colnames</span>(Hapi_input[[donor]]))</span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a>    <span class="fu">colnames</span>(Hapi_input[[donor]]) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"NFF"</span>, <span class="st">"NF"</span>, <span class="fu">colnames</span>(Hapi_input[[donor]]))</span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>    NA_count <span class="ot">&lt;-</span> <span class="fu">t</span>(Hapi_input[[donor]] <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="sc">-</span><span class="fu">c</span>(chr, pos, ref, alt)) <span class="sc">%&gt;%</span> <span class="fu">summarise_all</span>(<span class="sc">~</span><span class="fu">sum</span>(<span class="fu">is.na</span>(.))))<span class="sc">/</span><span class="fu">nrow</span>(Hapi_input[[donor]])</span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>    NA_count_chr <span class="ot">&lt;-</span> <span class="fu">rbind</span>(NA_count_chr, NA_count)</span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(Hapi_input, <span class="fu">paste0</span>(<span class="st">"../datasets/Hapi_input/QUAL_and_RGQ_aneu_removed/chr"</span>, chr_id, <span class="st">".rds"</span>))</span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colnames</span>(NA_count_chr) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"chr"</span>, chr_id)</span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-32"><a href="#cb2-32" aria-hidden="true" tabindex="-1"></a>  NA_count_chr_list[[<span class="fu">paste0</span>(<span class="st">"chr"</span>, chr_id)]] <span class="ot">&lt;-</span> NA_count_chr</span>
<span id="cb2-33"><a href="#cb2-33" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-34"><a href="#cb2-34" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-35"><a href="#cb2-35" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb2-36"><a href="#cb2-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-37"><a href="#cb2-37" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(NA_count_chr_list, <span class="fu">paste0</span>(<span class="st">"../datasets/Hapi_input/QUAL_and_RGQ_aneu_removed/NA_count_chr_list.rds"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
<section id="check-the-na-density" class="level2" data-number="4.5">
<h2 data-number="4.5" class="anchored" data-anchor-id="check-the-na-density"><span class="header-section-number">4.5</span> Check the NA density</h2>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R script</strong></pre>
</div>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>NA_count_chr_list <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"../datasets/Hapi_input/QUAL_and_RGQ_aneu_removed/NA_count_chr_list.rds"</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>NA_count_chr <span class="ot">&lt;-</span> <span class="fu">do.call</span>(cbind, NA_count_chr_list)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(NA_count_chr) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"N"</span>, <span class="st">"NF"</span>, <span class="fu">rownames</span>(NA_count_chr))</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(NA_count_chr) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"NFF"</span>, <span class="st">"NF"</span>, <span class="fu">rownames</span>(NA_count_chr))</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>NA_count_chr <span class="ot">&lt;-</span> NA_count_chr <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span> tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(<span class="st">"sperm_id"</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">Individual =</span> <span class="fu">gsub</span>(<span class="st">"s</span><span class="sc">\\</span><span class="st">d+"</span>, <span class="st">""</span>, sperm_id), <span class="at">Type =</span> <span class="fu">ifelse</span>(<span class="fu">grepl</span>(<span class="st">"AS"</span>, Individual), <span class="st">"AS"</span>, <span class="st">"NF"</span>))</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>NA_count_chr <span class="ot">&lt;-</span> <span class="fu">melt</span>(NA_count_chr, <span class="at">id=</span><span class="fu">c</span>(<span class="st">"sperm_id"</span>, <span class="st">"Individual"</span>, <span class="st">"Type"</span>))</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(NA_count_chr, <span class="fu">aes</span>(<span class="at">x=</span>value, <span class="at">color=</span>Type)) <span class="sc">+</span></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>() <span class="sc">+</span> <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values=</span><span class="fu">c</span>(<span class="st">"coral"</span>, <span class="st">"cornflowerblue"</span>)) <span class="sc">+</span> </span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>( variable <span class="sc">~</span> .)</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="fu">paste0</span>(<span class="st">"../figures/QUAL_and_RGQ_aneu_removed_NA_Density.pdf"</span>), <span class="at">width =</span> <span class="dv">8</span>, <span class="at">height =</span> <span class="dv">22</span><span class="sc">*</span><span class="dv">3</span>, <span class="at">limitsize =</span> <span class="cn">FALSE</span>)</span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>NA_count_chr <span class="ot">&lt;-</span> <span class="fu">do.call</span>(cbind, NA_count_chr_list)</span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(NA_count_chr) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"N"</span>, <span class="st">"NF"</span>, <span class="fu">rownames</span>(NA_count_chr))</span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(NA_count_chr) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"NFF"</span>, <span class="st">"NF"</span>, <span class="fu">rownames</span>(NA_count_chr))</span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a>NA_count_chr <span class="ot">&lt;-</span> NA_count_chr <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span> tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(<span class="st">"sperm_id"</span>)</span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a>NA_count_chr<span class="sc">$</span>larger_07_count <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(NA_count_chr[, <span class="sc">-</span><span class="dv">1</span>] <span class="sc">&gt;</span> <span class="fl">0.7</span>, <span class="at">na.rm=</span><span class="cn">FALSE</span>)</span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a>remove_by_NA_sperm_list <span class="ot">&lt;-</span> NA_count_chr<span class="sc">$</span>sperm_id[NA_count_chr<span class="sc">$</span>larger_07_count <span class="sc">&gt;</span> <span class="dv">2</span>]</span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(remove_by_NA_sperm_list, <span class="st">"../datasets/Hapi_input/QUAL_and_RGQ_aneu_removed/remove_by_NA_sperm_list.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
<section id="regenerate-the-hapi-input" class="level2" data-number="4.6">
<h2 data-number="4.6" class="anchored" data-anchor-id="regenerate-the-hapi-input"><span class="header-section-number">4.6</span> Regenerate the Hapi input</h2>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R script</strong></pre>
</div>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (chr_id <span class="cf">in</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">22</span>)) {</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>  Hapi_input <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(<span class="st">"../datasets/Hapi_input/QUAL_and_RGQ_aneu_removed/chr"</span>, chr_id, <span class="st">".rds"</span>))</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (donor <span class="cf">in</span> <span class="fu">names</span>(Hapi_input)) {</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">colnames</span>(Hapi_input[[donor]]) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"N"</span>, <span class="st">"NF"</span>, <span class="fu">colnames</span>(Hapi_input[[donor]]))</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">colnames</span>(Hapi_input[[donor]]) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"NFF"</span>, <span class="st">"NF"</span>, <span class="fu">colnames</span>(Hapi_input[[donor]]))</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>    Hapi_input[[donor]] <span class="ot">&lt;-</span> Hapi_input[[donor]][,<span class="sc">!</span><span class="fu">names</span>(Hapi_input[[donor]]) <span class="sc">%in%</span> remove_by_NA_sperm_list]</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(Hapi_input, <span class="fu">paste0</span>(<span class="st">"D:/sperm_project/Hapi_input/QUAL_and_RGQ_aneu_NA_removed_3/chr"</span>, chr_id, <span class="st">".rds"</span>))</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
<section id="count-the-sperm-number-in-each-individual-after-filter" class="level2" data-number="4.7">
<h2 data-number="4.7" class="anchored" data-anchor-id="count-the-sperm-number-in-each-individual-after-filter"><span class="header-section-number">4.7</span> Count the sperm number in each individual after filter</h2>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R script</strong></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>sperms_count <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>Hapi_input <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(<span class="st">"../datasets/Hapi_input/QUAL_and_RGQ_aneu_NA_removed_3/chr"</span>, <span class="dv">1</span>, <span class="st">".rds"</span>))</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (donor <span class="cf">in</span> <span class="fu">names</span>(Hapi_input)) {</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>    sperms_count <span class="ot">&lt;-</span> <span class="fu">c</span>(sperms_count, <span class="fu">ncol</span>(Hapi_input[[donor]]) <span class="sc">-</span> <span class="dv">4</span>)</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(sperms_count) <span class="ot">&lt;-</span> <span class="fu">names</span>(Hapi_input)</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>remove_donor <span class="ot">&lt;-</span> <span class="fu">names</span>(sperms_count[sperms_count <span class="sc">&lt;</span> <span class="dv">3</span>])</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(remove_donor, <span class="st">"../datasets/Hapi_input/QUAL_and_RGQ_aneu_NA_removed_3/remove_donor.rds"</span>)</span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (chr_id <span class="cf">in</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">22</span>)) {</span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a>  Hapi_input <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">paste0</span>(<span class="st">"../datasets/Hapi_input/QUAL_and_RGQ_aneu_NA_removed_3/chr"</span>, chr_id, <span class="st">".rds"</span>))</span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (donor <span class="cf">in</span> remove_donor) {</span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a>    Hapi_input[[donor]] <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">saveRDS</span>(Hapi_input, <span class="fu">paste0</span>(<span class="st">"../datasets/Hapi_input/QUAL_and_RGQ_aneu_NA_removed_3/chr"</span>, chr_id, <span class="st">".rds"</span>))</span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./section_2_aneuploidy_test.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Aneuploidy Test</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./section_4_run_Hapi.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Run HAPI</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>